### DISCUSSION TITLE
Install autokeras with mac-optimized tensorflow?

### DISCUSSION THREAD

**Discussion by geometrid** [keras-team/autokeras#1519] [Q&A]
Hi there! 👋  I'm relatively new to autokeras but I'm already a big fan, cheers!

I'm wondering: is it possible to install autokeras on macOS to take advantage of a [mac-optimized version of tensorflow](https://github.com/apple/tensorflow_macos)?

In the past, I've successfully installed autokeras with (CPU-only, non-optimized) tf2 on mac as well as with GPU tf2 on a linux system, but my first attempt at installing autokeras with mac-optimized tf2 on apple m1 didn't work. Happy to try again and provide more details, but wanted to post this question here in case this is obviously supported or obviously not supported. I searched through the existing questions but couldn't find anything relevant to this.

Thanks in advance for your help! 🙌

**Comment by haifeng-jin** [comment]
Such support is purely depending on Tensorflow.
If anyone had successfully run TF with mac optimized version.
AutoKeras should be able to run as well.

**Reply by geometrid** [reply]
Hi @haifeng-jin ! Yes, totally agree in theory-- but in this case, I have a successful mac-optimized TF install (as far as I can tell so far), but installing keras-tuner as part of the AutoKeras installation fails.

I have some experience installing TF + AK in other environments (CPU-only non-optimized on intel mac, and GPU-support on ubuntu), but I'm quite new to working with mac-optimized on Apple silicon / arm64, so I could be doing something obviously wrong, but 🤷‍♂️ . Following the AK installation steps, instead of installing TF via pip, I follow [this guide to using Apple's mac-optimized TF packages](https://github.com/apple/tensorflow_macos/issues/153), and although I haven't put it fully through its paces, this TF install seems successful; `tf2-arm64` is the name of the conda-managed venv:

```
(tf2-arm64) ~ % ipython
Python 3.8.8 | packaged by conda-forge | (default, Feb 20 2021, 15:50:57)
Type 'copyright', 'credits' or 'license' for more information
IPython 7.21.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import tensorflow as tf2

In [2]: hello = tf2.constant("hello TensorFlow!")

In [3]: print(hello)
tf.Tensor(b'hello TensorFlow!', shape=(), dtype=string)
```

However, continuing with the AK install instructions, I get 3.7k lines of errors when trying to pip install keras-tuner:

```
(tf2-arm64) ~ % pip install -U keras-tuner > pip_remote_install_keras_tuner.log 2> pip_remote_install_keras_tuner_err.log
(tf2-arm64) ~ % wc pip_remote_install_keras_tuner_err.log
    3701   21290  309141 pip_remote_install_keras_tuner_err.log
```

There's a lot of cruft in the error log, but I see a number of lines involving numpy:

```
(tf2-arm64) ~ % cat pip_remote_install_keras_tuner_err.log| grep numpy | wc
     707    4684  100667
(tf2-arm64) ~ % head -n 15 pip_remote_install_keras_tuner_err.log
...
  Ignoring numpy: markers 'python_version == "3.7" and platform_machine != "aarch64" and platform_system == "AIX"' don't match your environment
  Ignoring numpy: markers 'python_version == "3.6" and platform_machine == "aarch64"' don't match your environment
  Ignoring numpy: markers 'python_version == "3.7" and platform_machine == "aarch64"' don't match your environment
  Ignoring numpy: markers 'python_version == "3.8" and platform_machine == "aarch64"' don't match your environment
  Ignoring numpy: markers 'python_version == "3.7" and platform_machine != "aarch64" and platform_system != "AIX"' don't match your environment
  Ignoring numpy: markers 'python_version == "3.9"' don't match your environment
```

(FWIW I get similar errors when trying to install keras-tuner from source rather than remotely via pip.) I'm a bit out of my depth here-- hence this thread-- but at the moment I'm wondering if the numpy version (1.20.1) that was installed via the TF install outlined above is incompatible with keras-tuner and/or AK? I've looked at setup.py for each project and there doesn't seem to be any explicit numpy version requirement. And as of the time of this post, only a few numpy versions are [available for arm64 on conda-forge](https://conda.anaconda.org/conda-forge/osx-arm64/): 1.19.2, 1.19.4, 1.19.5, 1.20.0, and 1.20.1. Perhaps there are other conda or pip channels with other arm64 numpy versions? Or perhaps I'm barking up the wrong tree altogether. Maybe it's a pip issue? Or something else? Again, 🤷‍♂️ 

In any case, thanks again! (For reference, I also raised this [issue on the keras-tuner github](https://github.com/keras-team/keras-tuner/issues/493).)

**Reply by haifeng-jin** [reply]
Sorry, I have no idea.
I do not have the hardware to reproduce the error.
Does anyone have an idea?

**Comment by stancsz** [comment]
I believe it is a tensorflow/apple hardware compatibility issue, have you tried [searching issues in here?](https://github.com/apple/tensorflow_macos)

can you also provide? 
```
$platform_machine
$platform_system
```

**Reply by geometrid** [reply]
Hi there! Thanks for your help. You might be right, but I followed the mac-optimized tensorflow install instructions and I have a working mac-optimized tensorflow installed, so I don't think it's a tf/mac issue. Another user suggested a workaround [in issue #1466 ](https://github.com/keras-team/autokeras/issues/1466#issuecomment-812314281), which seems promising, but I haven't verified that it works on my environment yet. But step (5) in that comment is a new idea to me, so I'm hopeful that it will work for me.

For reference though:

```
In [1]: import platform

In [2]: platform.machine()
Out[2]: 'arm64'

In [3]: platform.system()
Out[3]: 'Darwin'
```

Thanks!

**Comment by Python-Ninja-Hebi** [comment]
I installed tensorflow like described here https://developer.apple.com/metal/tensorflow-plugin/

Then installed autokeras with pip but there are always error.

Was anyone successful with apple m1 and autokeras?

**Reply by thomascheuk** [reply]
There is a workaround [here](https://github.com/keras-team/autokeras/issues/1466#issuecomment-812314281)
